// @toruslabs/ethereum-controllers/dist/lib.esm/Eip5792/walletGetCallsStatus.js
import { addHexPrefix } from '@ethereumjs/util';
import { TransactionStatus } from '@toruslabs/base-controllers';
import { rpcErrors, JsonRpcError } from '@web3auth/auth';
import { GetCallsStatusCode, EIP_5792_METHODS, EIP5792ErrorCode } from '../utils/eip5792Types.js';

/**
 * Maps a TransactionStatus to an EIP-5792 GetCallsStatusCode.
 * @param status - The transaction status.
 * @returns The corresponding EIP-5792 status code.
 */
function mapTransactionStatusToEip5792Status(status) {
  switch (status) {
    // Pending states - batch is still being processed
    case TransactionStatus.unapproved:
    case TransactionStatus.approved:
    case TransactionStatus.signed:
    case TransactionStatus.submitted:
    case TransactionStatus.pending:
    case TransactionStatus.cancelling:
      return GetCallsStatusCode.PENDING;

    // Confirmed states - batch completed successfully
    case TransactionStatus.confirmed:
    case TransactionStatus.finalized:
    case TransactionStatus.processed:
      return GetCallsStatusCode.CONFIRMED;

    // Failed offchain - rejected/cancelled before reaching the chain
    case TransactionStatus.rejected:
    case TransactionStatus.cancelled:
    case TransactionStatus.expired:
    case TransactionStatus.dropped:
      return GetCallsStatusCode.FAILED_OFFCHAIN;

    // Reverted - transaction was submitted but failed on chain
    case TransactionStatus.failed:
      return GetCallsStatusCode.REVERTED;
    default:
      return GetCallsStatusCode.PENDING;
  }
}

/**
 * Handler for wallet_getCallsStatus (EIP-5792).
 * Returns the status of a batch of calls.
 *
 * @param request - The JRPC request with batch ID parameter.
 * @param context - Context containing required functions.
 * @returns The batch status and receipts.
 */
function walletGetCallsStatus(request, getTransactionByBatchId) {
  const {
    method,
    params
  } = request;
  if (method !== EIP_5792_METHODS.WALLET_GET_CALLS_STATUS) {
    return;
  }
  const batchId = Array.isArray(params) ? params[0] : params;
  if (!batchId) {
    throw rpcErrors.invalidParams("wallet_getCallsStatus: batch ID is required");
  }
  const batchTx = getTransactionByBatchId(batchId);
  if (!batchTx) {
    throw new JsonRpcError(EIP5792ErrorCode.UnknownBundleId, `wallet_getCallsStatus: Batch not found: ${batchId}`);
  }
  return {
    status: mapTransactionStatusToEip5792Status(batchTx.status),
    id: batchId,
    chainId: addHexPrefix(batchTx.chainId),
    atomic: true,
    receipts: batchTx.txReceipt ? [batchTx.txReceipt] : []
  };
}

export { mapTransactionStatusToEip5792Status, walletGetCallsStatus };
