// @toruslabs/ethereum-controllers/dist/lib.esm/Eip5792/walletSendCalls.js
import { isValidAddress } from '@ethereumjs/util';
import { rpcErrors } from '@web3auth/auth';
import { isHexString } from 'ethers';
import { v4, parse } from 'uuid';
import { getIsEip7702UpgradeSupported } from '../Eip7702/eip7702Utils.js';
import { EIP_5792_METHODS } from '../utils/eip5792Types.js';

/**
 * Generates a unique batch ID for EIP-5792 calls.
 * @returns A unique hex string batch ID.
 */
function generateBatchId() {
  const idString = v4();
  const idBytes = new Uint8Array(parse(idString));
  const hexString = Array.from(idBytes).map(byte => byte.toString(16).padStart(2, "0")).join("");
  return `0x${hexString}`;
}

/**
 * Validates a single call in the batch.
 * @param call - The call to validate.
 * @param index - The index of the call in the batch (for error messages).
 */
function validateCall(call, index) {
  // Validate 'to' address
  if (!call.to || !isHexString(call.to) || !isValidAddress(call.to)) {
    throw rpcErrors.invalidParams(`Invalid 'to' address in call at index ${index}`);
  }

  // Validate 'value' if present
  if (call.value !== undefined) {
    if (!isHexString(call.value)) {
      throw rpcErrors.invalidParams(`Invalid 'value' in call at index ${index}: must be a valid non-negative hex string`);
    }
  }

  // Validate 'data' if present
  if (call.data !== undefined) {
    if (!isHexString(call.data)) {
      throw rpcErrors.invalidParams(`Invalid 'data' in call at index ${index}: must be a valid hex string`);
    }
  }
}

/**
 * Validates the parameters for wallet_sendCalls (EIP-5792).
 * @param sendCallsParams - The parameters to validate.
 */
function validateSendCallsParams(sendCallsParams) {
  // Basic structure validation
  if (!sendCallsParams) {
    throw rpcErrors.invalidParams("Missing send calls parameters");
  }

  // Validate version format
  if (!sendCallsParams.version || typeof sendCallsParams.version !== "string") {
    throw rpcErrors.invalidParams(`Invalid version: expected a string, got "${sendCallsParams.version || "undefined"}"`);
  }

  // Validate chainId format
  if (!sendCallsParams.chainId || !isHexString(sendCallsParams.chainId)) {
    throw rpcErrors.invalidParams("Invalid chainId: must be a valid hex string");
  }

  // Validate 'from' address
  if (!sendCallsParams.from) {
    throw rpcErrors.invalidParams("Missing 'from' address");
  }
  if (!isHexString(sendCallsParams.from) || !isValidAddress(sendCallsParams.from)) {
    throw rpcErrors.invalidParams("Invalid 'from' address: must be a valid hex address");
  }

  // Validate calls array
  if (!Array.isArray(sendCallsParams.calls) || sendCallsParams.calls.length === 0) {
    throw rpcErrors.invalidParams("Invalid calls: must be a non-empty array");
  }

  // Validate each call
  sendCallsParams.calls.forEach((call, index) => {
    validateCall(call, index);
  });
}
async function processMultipleTransactions({
  from,
  transactions,
  request,
  chainId,
  getEthCode,
  processTransactionBatch
}) {
  // check if the account is upgraded with EIP-7702 for the given chain
  const {
    isSupported,
    upgradeContractAddress,
    delegationAddress
  } = await getIsEip7702UpgradeSupported(from, chainId, getEthCode);
  if (!isSupported) {
    throw rpcErrors.methodNotSupported(`EIP-7702 upgrade is not supported for the given chain, ${chainId}`);
  }
  const hasAccountAlreadyUpgraded = Boolean(delegationAddress);
  const batchRequest = {
    batchId: generateBatchId(),
    transactions,
    eip7702UpgradeContractAddress: upgradeContractAddress,
    requiredEip7702Upgrade: !hasAccountAlreadyUpgraded
  };

  // prepare batch transactions
  return processTransactionBatch(batchRequest, request);
}
async function processSingleTransaction({
  transactions,
  request,
  processTransactionBatch
}) {
  const batchRequestWithSingleCall = {
    batchId: generateBatchId(),
    transactions
  };
  return processTransactionBatch(batchRequestWithSingleCall, request);
}

/**
 * Handler for wallet_sendCalls (EIP-5792).
 * Sends a batch of calls for an EIP-7702 upgraded account.
 *
 * @param request - The JRPC request with send calls parameters.
 * @param getEthCode - Function to get the code at an address.
 * @param context - Context containing required functions.
 * @returns The batch ID.
 */
async function walletSendCalls(request, getEthCode, context) {
  const {
    method,
    params
  } = request;
  if (method !== EIP_5792_METHODS.WALLET_SEND_CALLS) {
    return;
  }
  const sendCallsParams = Array.isArray(params) ? params[0] : params;
  validateSendCallsParams(sendCallsParams);
  const {
    chainId,
    from,
    calls
  } = sendCallsParams;
  const transactions = calls.map(call => ({
    params: call
  }));
  const isBatchCalls = Object.keys(transactions).length > 1;
  if (!isBatchCalls) {
    return processSingleTransaction({
      transactions,
      request,
      processTransactionBatch: context.processTransactionBatch
    });
  }
  return processMultipleTransactions({
    from: from,
    transactions,
    request,
    chainId,
    getEthCode,
    processTransactionBatch: context.processTransactionBatch
  });
}

export { generateBatchId, processMultipleTransactions, processSingleTransaction, validateSendCallsParams, walletSendCalls };
