// @toruslabs/ethereum-controllers/dist/lib.cjs/utils/eip5792Types.js
'use strict';

// ============================================================================
// EIP-5792 Method Names
// ============================================================================
const EIP_5792_METHODS = {
  WALLET_GET_CAPABILITIES: "wallet_getCapabilities",
  WALLET_SEND_CALLS: "wallet_sendCalls",
  WALLET_GET_CALLS_STATUS: "wallet_getCallsStatus",
  WALLET_SHOW_CALLS_STATUS: "wallet_showCallsStatus"
};
const EIP5792ErrorCode = {
  UnsupportedNonOptionalCapability: 5700,
  UnsupportedChainId: 5710,
  UnknownBundleId: 5730,
  RejectedUpgrade: 5750
};
const GetCallsStatusCode = {
  PENDING: 100,
  CONFIRMED: 200,
  FAILED_OFFCHAIN: 400,
  REVERTED: 500,
  REVERTED_PARTIAL: 600
};
/**
 * EIP-5792 atomic status (EIP-5792)
 *
 * Reference: https://eips.ethereum.org/EIPS/eip-5792#wallet_getcapabilities
 */
const Eip5792AtomicStatus = {
  SUPPORTED: "supported",
  READY: "ready",
  UNSUPPORTED: "unsupported"
};

exports.EIP5792ErrorCode = EIP5792ErrorCode;
exports.EIP_5792_METHODS = EIP_5792_METHODS;
exports.Eip5792AtomicStatus = Eip5792AtomicStatus;
exports.GetCallsStatusCode = GetCallsStatusCode;
