// @toruslabs/ethereum-controllers/dist/lib.esm/Eip7702/eip7702Utils.js
import { addHexPrefix } from '@ethereumjs/util';
import { zeroAddress, encodeAbiParameters, parseAbiParameters, encodeFunctionData } from 'viem';
import { erc7821Abi } from '../utils/abis.js';
import { SUPPORTED_NETWORKS } from '../utils/constants.js';
import { EIP_7702_PREFIX } from '../utils/eip7702Types.js';

/**
 * The MetaMask EIP-7702 Stateless Delegator contract address.
 * Used as the delegation target for all supported networks.
 */
const MetaMask_EIP7702_Stateless_Delegator = "0x63c0c19a282a1B52b07dD5a65b58948A07DAE32B";
async function getDelegationAddress(walletAddress, chainId, getEthCode) {
  // query eth_getCode
  const code = await getEthCode(walletAddress, chainId);
  const normalizedCode = code ? addHexPrefix(code).toLowerCase() : "0x0";
  const hasDelegation = normalizedCode.length === 48 && normalizedCode.startsWith(EIP_7702_PREFIX);
  if (hasDelegation) {
    const delegationAddress = addHexPrefix(normalizedCode.slice(EIP_7702_PREFIX.length));
    return delegationAddress;
  }
  return null;
}

/**
 * Check if the wallet is supported for EIP-7702 on provided chain.
 */
async function getIsEip7702UpgradeSupported(address, chainId, getEthCode) {
  if (!SUPPORTED_NETWORKS[chainId]) {
    return {
      isSupported: false,
      delegationAddress: null
    };
  }
  const delegationAddress = await getDelegationAddress(address, chainId, getEthCode);
  return {
    isSupported: true,
    upgradeContractAddress: MetaMask_EIP7702_Stateless_Delegator,
    delegationAddress
  };
}

/**
 * Generate an EIP-7702 batch transaction.
 *
 * Reference: {@link https://github.com/MetaMask/core/blob/main/packages/transaction-controller/src/utils/eip7702.ts#L119}
 *
 * @param from - The sender address.
 * @param transactions - The transactions to batch.
 * @returns The batch transaction.
 */
function generateEIP7702BatchTransaction(from, transactions) {
  const calls = transactions.map(transaction => {
    const {
      data,
      to,
      value
    } = transaction;
    return {
      target: to !== null && to !== void 0 ? to : zeroAddress,
      value: BigInt(value !== null && value !== void 0 ? value : "0x0"),
      callData: data !== null && data !== void 0 ? data : "0x"
    };
  });

  // Single batch mode, no opData.
  const mode = "0x01".padEnd(66, "0");

  // Encode the calls array as (address target, uint256 value, bytes callData)[]
  const executionData = encodeAbiParameters(parseAbiParameters("(address target, uint256 value, bytes callData)[]"), [calls]);
  const data = encodeFunctionData({
    abi: erc7821Abi,
    functionName: "execute",
    args: [mode, executionData]
  });
  return {
    data,
    to: from
  };
}

export { MetaMask_EIP7702_Stateless_Delegator, generateEIP7702BatchTransaction, getDelegationAddress, getIsEip7702UpgradeSupported };
