// @toruslabs/ethereum-controllers/dist/lib.esm/Eip7702/walletUpgradeAccount.js
import { rpcErrors } from '@web3auth/auth';
import { TRANSACTION_ENVELOPE_TYPES } from '../utils/constants.js';
import { getIsEip7702UpgradeSupported } from './eip7702Utils.js';

/**
 * Handle and prepare the transaction to upgrade an EOA to the EIP-7702 smart account.
 *
 * @param req - The request object.
 * @param getEthCode - The function to get the code of an account.
 * @param processTransaction - The function to process a transaction.
 * @returns The transaction hash.
 */
async function walletUpgradeAccount(req, getEthCode, processTransaction) {
  const {
    method,
    params
  } = req;
  if (method !== "wallet_upgradeAccount") {
    return;
  }
  const upgradeAccountParam = Array.isArray(params) ? params[0] : params;
  const account = upgradeAccountParam.account;
  const chainId = upgradeAccountParam.chainId;
  if (!account || !chainId) {
    throw rpcErrors.invalidParams("wallet_upgradeAccount: account and chainId are required");
  }
  const {
    isSupported,
    upgradeContractAddress,
    delegationAddress
  } = await getIsEip7702UpgradeSupported(account, chainId, getEthCode);
  if (!isSupported) {
    throw rpcErrors.methodNotSupported(`Account upgrade is not supported for chain ${chainId}`);
  }
  if (!upgradeContractAddress) {
    throw rpcErrors.internal({
      message: `Upgrade contract address not found for chain ${chainId}`
    });
  }
  const isAlreadyUpgraded = Boolean(delegationAddress);
  if (isAlreadyUpgraded) {
    throw rpcErrors.invalidParams(`Account is already upgraded for chain ${chainId}`);
  }
  return processTransaction({
    from: account,
    to: account,
    type: TRANSACTION_ENVELOPE_TYPES.SET_CODE,
    // Type 4 transaction
    authorizationList: [{
      address: upgradeContractAddress,
      chainId
    }]
  }, req);
}

export { walletUpgradeAccount };
