// @toruslabs/ethereum-controllers/dist/lib.esm/utils/eip5792Types.js
// ============================================================================
// EIP-5792 Method Names
// ============================================================================

const EIP_5792_METHODS = {
  WALLET_GET_CAPABILITIES: "wallet_getCapabilities",
  WALLET_SEND_CALLS: "wallet_sendCalls",
  WALLET_GET_CALLS_STATUS: "wallet_getCallsStatus",
  WALLET_SHOW_CALLS_STATUS: "wallet_showCallsStatus"
};

// ============================================================================
// EIP-5792 Types
// ============================================================================

/**
 * A single call in a batch (EIP-5792)
 */

/**
 * Parameters for wallet_sendCalls (EIP-5792)
 */

/**
 * Parameters for wallet_getCallsStatus (EIP-5792)
 */
// batch ID

/**
 * Parameters for wallet_showCallsStatus (EIP-5792)
 */
// batch ID

/**
 * Parameters for wallet_getCapabilities (EIP-5792)
 */

// [wallet address, supported chains (optional)]

const EIP5792ErrorCode = {
  UnsupportedNonOptionalCapability: 5700,
  UnsupportedChainId: 5710,
  UnknownBundleId: 5730,
  RejectedUpgrade: 5750
};
const GetCallsStatusCode = {
  PENDING: 100,
  CONFIRMED: 200,
  FAILED_OFFCHAIN: 400,
  REVERTED: 500,
  REVERTED_PARTIAL: 600
};

/**
 * Response for wallet_getCallsStatus (EIP-5792)
 */

/**
 * EIP-5792 atomic status (EIP-5792)
 *
 * Reference: https://eips.ethereum.org/EIPS/eip-5792#wallet_getcapabilities
 */
const Eip5792AtomicStatus = {
  SUPPORTED: "supported",
  READY: "ready",
  UNSUPPORTED: "unsupported"
};

/**
 * Capabilities for a chain (EIP-5792)
 */

/**
 * Response for wallet_getCapabilities (EIP-5792)
 */

/**
 * Internal representation of a batch of calls
 */

/**
 * Function type for getting supported chains
 */

/**
 * Function type for getting cached delegations
 */

// ============================================================================
// Batch Transaction Types
// ============================================================================

/**
 * The parameters of a transaction within an atomic batch.
 */

/** Metadata for a nested transaction within a standard transaction. */

/**
 * Request to submit a batch of transactions.
 * Currently only atomic batches are supported via EIP-7702.
 */

export { EIP5792ErrorCode, EIP_5792_METHODS, Eip5792AtomicStatus, GetCallsStatusCode };
