// @toruslabs/ethereum-controllers/dist/lib.cjs/Eip5792/walletGetCallsStatus.js
'use strict';

var util = require('@ethereumjs/util');
var baseControllers = require('@toruslabs/base-controllers');
var auth = require('@web3auth/auth');
var eip5792Types = require('../utils/eip5792Types.js');

/**
 * Maps a TransactionStatus to an EIP-5792 GetCallsStatusCode.
 * @param status - The transaction status.
 * @returns The corresponding EIP-5792 status code.
 */
function mapTransactionStatusToEip5792Status(status) {
  switch (status) {
    // Pending states - batch is still being processed
    case baseControllers.TransactionStatus.unapproved:
    case baseControllers.TransactionStatus.approved:
    case baseControllers.TransactionStatus.signed:
    case baseControllers.TransactionStatus.submitted:
    case baseControllers.TransactionStatus.pending:
    case baseControllers.TransactionStatus.cancelling:
      return eip5792Types.GetCallsStatusCode.PENDING;
    // Confirmed states - batch completed successfully
    case baseControllers.TransactionStatus.confirmed:
    case baseControllers.TransactionStatus.finalized:
    case baseControllers.TransactionStatus.processed:
      return eip5792Types.GetCallsStatusCode.CONFIRMED;
    // Failed offchain - rejected/cancelled before reaching the chain
    case baseControllers.TransactionStatus.rejected:
    case baseControllers.TransactionStatus.cancelled:
    case baseControllers.TransactionStatus.expired:
    case baseControllers.TransactionStatus.dropped:
      return eip5792Types.GetCallsStatusCode.FAILED_OFFCHAIN;
    // Reverted - transaction was submitted but failed on chain
    case baseControllers.TransactionStatus.failed:
      return eip5792Types.GetCallsStatusCode.REVERTED;
    default:
      return eip5792Types.GetCallsStatusCode.PENDING;
  }
}
/**
 * Handler for wallet_getCallsStatus (EIP-5792).
 * Returns the status of a batch of calls.
 *
 * @param request - The JRPC request with batch ID parameter.
 * @param context - Context containing required functions.
 * @returns The batch status and receipts.
 */
function walletGetCallsStatus(request, getTransactionByBatchId) {
  const {
    method,
    params
  } = request;
  if (method !== eip5792Types.EIP_5792_METHODS.WALLET_GET_CALLS_STATUS) {
    return;
  }
  const batchId = Array.isArray(params) ? params[0] : params;
  if (!batchId) {
    throw auth.rpcErrors.invalidParams("wallet_getCallsStatus: batch ID is required");
  }
  const batchTx = getTransactionByBatchId(batchId);
  if (!batchTx) {
    throw new auth.JsonRpcError(eip5792Types.EIP5792ErrorCode.UnknownBundleId, `wallet_getCallsStatus: Batch not found: ${batchId}`);
  }
  return {
    status: mapTransactionStatusToEip5792Status(batchTx.status),
    id: batchId,
    chainId: util.addHexPrefix(batchTx.chainId),
    atomic: true,
    receipts: batchTx.txReceipt ? [batchTx.txReceipt] : []
  };
}

exports.mapTransactionStatusToEip5792Status = mapTransactionStatusToEip5792Status;
exports.walletGetCallsStatus = walletGetCallsStatus;
