// @toruslabs/ethereum-controllers/dist/lib.cjs/Eip7702/eip7702Utils.js
'use strict';

var util = require('@ethereumjs/util');
var viem = require('viem');
var abis = require('../utils/abis.js');
var constants = require('../utils/constants.js');
var eip7702Types = require('../utils/eip7702Types.js');

/**
 * The MetaMask EIP-7702 Stateless Delegator contract address.
 * Used as the delegation target for all supported networks.
 */
const MetaMask_EIP7702_Stateless_Delegator = "0x63c0c19a282a1B52b07dD5a65b58948A07DAE32B";
async function getDelegationAddress(walletAddress, chainId, getEthCode) {
  // query eth_getCode
  const code = await getEthCode(walletAddress, chainId);
  const normalizedCode = code ? util.addHexPrefix(code).toLowerCase() : "0x0";
  const hasDelegation = normalizedCode.length === 48 && normalizedCode.startsWith(eip7702Types.EIP_7702_PREFIX);
  if (hasDelegation) {
    const delegationAddress = util.addHexPrefix(normalizedCode.slice(eip7702Types.EIP_7702_PREFIX.length));
    return delegationAddress;
  }
  return null;
}
/**
 * Check if the wallet is supported for EIP-7702 on provided chain.
 */
async function getIsEip7702UpgradeSupported(address, chainId, getEthCode) {
  if (!constants.SUPPORTED_NETWORKS[chainId]) {
    return {
      isSupported: false,
      delegationAddress: null
    };
  }
  const delegationAddress = await getDelegationAddress(address, chainId, getEthCode);
  return {
    isSupported: true,
    upgradeContractAddress: MetaMask_EIP7702_Stateless_Delegator,
    delegationAddress
  };
}
/**
 * Generate an EIP-7702 batch transaction.
 *
 * Reference: {@link https://github.com/MetaMask/core/blob/main/packages/transaction-controller/src/utils/eip7702.ts#L119}
 *
 * @param from - The sender address.
 * @param transactions - The transactions to batch.
 * @returns The batch transaction.
 */
function generateEIP7702BatchTransaction(from, transactions) {
  const calls = transactions.map(transaction => {
    const {
      data,
      to,
      value
    } = transaction;
    return {
      target: to !== null && to !== void 0 ? to : viem.zeroAddress,
      value: BigInt(value !== null && value !== void 0 ? value : "0x0"),
      callData: data !== null && data !== void 0 ? data : "0x"
    };
  });
  // Single batch mode, no opData.
  const mode = "0x01".padEnd(66, "0");
  // Encode the calls array as (address target, uint256 value, bytes callData)[]
  const executionData = viem.encodeAbiParameters(viem.parseAbiParameters("(address target, uint256 value, bytes callData)[]"), [calls]);
  const data = viem.encodeFunctionData({
    abi: abis.erc7821Abi,
    functionName: "execute",
    args: [mode, executionData]
  });
  return {
    data,
    to: from
  };
}

exports.MetaMask_EIP7702_Stateless_Delegator = MetaMask_EIP7702_Stateless_Delegator;
exports.generateEIP7702BatchTransaction = generateEIP7702BatchTransaction;
exports.getDelegationAddress = getDelegationAddress;
exports.getIsEip7702UpgradeSupported = getIsEip7702UpgradeSupported;
