// @toruslabs/ethereum-controllers/dist/lib.cjs/Eip7702/walletUpgradeAccount.js
'use strict';

var auth = require('@web3auth/auth');
var constants = require('../utils/constants.js');
var eip7702Utils = require('./eip7702Utils.js');

/**
 * Handle and prepare the transaction to upgrade an EOA to the EIP-7702 smart account.
 *
 * @param req - The request object.
 * @param getEthCode - The function to get the code of an account.
 * @param processTransaction - The function to process a transaction.
 * @returns The transaction hash.
 */
async function walletUpgradeAccount(req, getEthCode, processTransaction) {
  const {
    method,
    params
  } = req;
  if (method !== "wallet_upgradeAccount") {
    return;
  }
  const upgradeAccountParam = Array.isArray(params) ? params[0] : params;
  const account = upgradeAccountParam.account;
  const chainId = upgradeAccountParam.chainId;
  if (!account || !chainId) {
    throw auth.rpcErrors.invalidParams("wallet_upgradeAccount: account and chainId are required");
  }
  const {
    isSupported,
    upgradeContractAddress,
    delegationAddress
  } = await eip7702Utils.getIsEip7702UpgradeSupported(account, chainId, getEthCode);
  if (!isSupported) {
    throw auth.rpcErrors.methodNotSupported(`Account upgrade is not supported for chain ${chainId}`);
  }
  if (!upgradeContractAddress) {
    throw auth.rpcErrors.internal({
      message: `Upgrade contract address not found for chain ${chainId}`
    });
  }
  const isAlreadyUpgraded = Boolean(delegationAddress);
  if (isAlreadyUpgraded) {
    throw auth.rpcErrors.invalidParams(`Account is already upgraded for chain ${chainId}`);
  }
  return processTransaction({
    from: account,
    to: account,
    type: constants.TRANSACTION_ENVELOPE_TYPES.SET_CODE,
    // Type 4 transaction
    authorizationList: [{
      address: upgradeContractAddress,
      chainId
    }]
  }, req);
}

exports.walletUpgradeAccount = walletUpgradeAccount;
